core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -672,24 +681,42 @@ public String toString()
         // This should not be a ConcurrentHashMap because we want to prevent concurrent addition of new consumers during the
         // removal of existing consumers from this map in addDynamicFilters. This ensures that new consumers don't miss filter completion.
         private final Map<DynamicFilterId, List<Consumer<Map<DynamicFilterId, Domain>>>> dynamicFilterConsumers = new HashMap<>();
-        private final long queryStartTime = System.nanoTime();
+        private long queryAttemptStartTime = System.nanoTime();
 
         private DynamicFilterContext(
                 Session session,
                 Set<DynamicFilterId> dynamicFilters,
-                Map<DynamicFilterId, SettableFuture<Void>> lazyDynamicFilters,
+                Set<DynamicFilterId> lazyDynamicFilters,
```

Comment:
More generally speaking I wonder if there's any real reason of not making the `DynamicFilterContext` interface `synchronized`. It doesn't seem to be highly contended and the operations are computationally simple. It would greatly simplify reasoning about thread safety of this class. Also it isn't clear whether having all data structure to be thread safe is even more efficient than having a single global lock + non thread safe data structure in a non heavily contended environment.

Reply:
We did start with simpler synchronisation in DynamicFilterService. I'm not sure what necessitated avoiding `synchronized` on DynamicFilterContext. @sopel39 do you remember why ?

Reply:
We wanted to avoid locking which could increase query latency. Implementation evolved over time so I'm not sure it matters that much now, but since there are no removals really (only additions), lock free implementation is rather easy.
Keep in mind that this is the place where all queries go, so it has potentials for generating lock conjestions.

> t doesn't seem to be highly contended and the operations are computationally simple.

They are not necessarily simple. We merge/union DFs together which can be substantial.

Reply:
I was under assumption that `DynamicFilterContext` is per query, isn't it the case?

I wonder if the  `merge/union` operation is the only "heavy" operation would it be possible to only perform this operation in a lock free way?

Also I'm thinking if it would be possible to even further reduce lock contention and get it to the "per-dynamic-filter" level.

e.g.: by structuring the `DynamicFilterContext` in a way that has a `Map<DynamicFilterId, DynamicFilterState>` where each `DynamicFilterState` has it's own lock.

Reply:
> I was under assumption that DynamicFilterContext is per query, isn't it the case?

Yes, it's per query

> Also I'm thinking if it would be possible to even further reduce lock contention and get it to the "per-dynamic-filter" level.

It is lock free on DF level.

> e.g.: by structuring the DynamicFilterContext in a way that has a Map<DynamicFilterId, DynamicFilterState> where each DynamicFilterState has it's own lock.

There are no locks currently. I'm not sure alternative approach with `synchronization` would be cleaner/simpler.